crates/ruff_server/src/server.rs

```diff
@@ -0,0 +1,152 @@
+//! Scheduling, I/O, and API endpoints.
+
+use anyhow::anyhow;
+use lsp::Connection;
+use lsp_server as lsp;
+use lsp_types as types;
+use types::ClientCapabilities;
+use types::CodeActionKind;
+use types::CodeActionOptions;
+use types::DiagnosticOptions;
+use types::OneOf;
+use types::PositionEncodingKind;
+use types::TextDocumentSyncCapability;
+use types::TextDocumentSyncKind;
+use types::TextDocumentSyncOptions;
+use types::WorkDoneProgressOptions;
+use types::WorkspaceFoldersServerCapabilities;
+
+use self::schedule::main_thread;
+use crate::session::Session;
+
+mod api;
+mod client;
+mod schedule;
+
+pub(crate) type Result<T> = std::result::Result<T, api::Error>;
+
+pub struct Server {
+    conn: lsp::Connection,
+    threads: lsp::IoThreads,
+    session: Session,
+}
+
+impl Server {
+    pub fn new() -> crate::Result<Self> {
+        let (conn, threads) = lsp::Connection::stdio();
+
+        let (id, params) = conn.initialize_start()?;
+
+        let init_params: types::InitializeParams = serde_json::from_value(params)?;
+
+        let client_capabilities = init_params.capabilities;
+        let server_capabilities = Self::server_capabilities(&client_capabilities);
+
+        let workspaces = init_params
+            .workspace_folders
+            .map(|f| f.into_iter().map(|w| w.uri).collect())
+            .or_else(|| init_params.root_uri.map(|u| vec![u])).ok_or_else(|| anyhow!("No workspace or root URI was given in the LSP initialization parameters. The server cannot start."))?;
```

Comment:
This is a very long line. Can you break it up? You can backslash escape newlines to break it up a bit: https://doc.rust-lang.org/rust-by-example/std/str.html#literals-and-escapes